src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_CollectionExpression.cs

```diff
@@ -231,101 +216,163 @@ private static bool CanOptimizeSingleSpreadAsCollectionBuilderArgument(BoundColl
             return spreadExpression is not null;
         }
 
-        private BoundExpression VisitImmutableArrayCollectionExpression(BoundCollectionExpression node, TypeWithAnnotations elementType, MethodSymbol asImmutableArray)
-        {
-            var arrayCreation = VisitArrayOrSpanCollectionExpression(
-                node,
-                CollectionExpressionTypeKind.Array,
-                ArrayTypeSymbol.CreateSZArray(_compilation.Assembly, elementType),
-                elementType);
-            // ImmutableCollectionsMarshal.AsImmutableArray(arrayCreation)
-            return _factory.StaticCall(asImmutableArray.Construct(ImmutableArray.Create(elementType)), ImmutableArray.Create(arrayCreation));
-        }
-
-        private BoundExpression VisitArrayOrSpanCollectionExpression(BoundCollectionExpression node, CollectionExpressionTypeKind collectionTypeKind, TypeSymbol collectionType, TypeWithAnnotations elementType)
+        /// <summary>
+        /// Create a collection value which may involve creating an array.
+        /// Handles types 'T[]', 'Span{T}', 'ReadOnlySpan{T}', 'ImmutableArray{T}'.
+        /// Note that Span/ROS/ImmutableArray cases, may or may not involve us actually creating an array.
+        /// Depending on the collection elements, available APIs, or other context, we may use some other strategy.
+        /// </summary>
+        private BoundExpression VisitArrayOrSpanCollectionExpression(BoundCollectionExpression node, TypeSymbol collectionType)
         {
             Debug.Assert(!_inExpressionLambda);
             Debug.Assert(_additionalLocals is { });
             Debug.Assert(node.CollectionCreation is null); // shouldn't have generated a constructor call
             Debug.Assert(node.Placeholder is null);
 
-            var syntax = node.Syntax;
-            var elements = node.Elements;
-            MethodSymbol? spanConstructor = null;
+            return collectionType switch
+            {
+                ArrayTypeSymbol arrayType
+                    => createArray(node, arrayType, targetsReadOnlyCollection: false),
+                NamedTypeSymbol spanType when spanType.OriginalDefinition.Equals(_compilation.GetWellKnownType(WellKnownType.System_Span_T))
+                    => createSpan(node, spanType, isReadOnlySpan: false),
+                NamedTypeSymbol readOnlySpanType when readOnlySpanType.OriginalDefinition.Equals(_compilation.GetWellKnownType(WellKnownType.System_ReadOnlySpan_T))
+                    => createSpan(node, readOnlySpanType, isReadOnlySpan: true),
+                NamedTypeSymbol immutableArrayType when immutableArrayType.OriginalDefinition.Equals(_compilation.GetWellKnownType(WellKnownType.System_Collections_Immutable_ImmutableArray_T))
+                    => createImmutableArray(node, immutableArrayType),
```

Comment:
this pattern was honestly odd.  switching three times as NamedTypeSYmbol each with a 'when' clause.  i don't particularly like it.  but not blocking it.